File: 36_Flip_Flop/javascript/flipflop.js
// FLIPFLOP
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var as = [];

// Main program
async function main()
{
    print(tab(32) + "FLIPFLOP\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    // *** Created by Michael Cass
    print("THE OBJECT OF THIS PUZZLE IS TO CHANGE THIS:\n");
    print("\n");
    print("X X X X X X X X X X\n");
    print("\n");
    print("TO THIS:\n");
    print("\n");
    print("O O O O O O O O O O\n");
    print("\n");
    print("BY TYPING THE NUMBER CORRESPONDING TO THE POSITION OF THE\n");
    print("LETTER ON SOME NUMBERS, ONE POSITION WILL CHANGE, ON\n");
    print("OTHERS, TWO WILL CHANGE.  TO RESET LINE TO ALL X'S, TYPE 0\n");
    print("(ZERO) AND TO START OVER IN THE MIDDLE OF A GAME, TYPE \n");
    print("11 (ELEVEN).\n");
    print("\n");
    while (1) {
        start = 1;
        do {
            z = 1;
            if (start == 1) {
                m = 0;
                q = Math.random();
                print("HERE IS THE STARTING LINE OF X'S.\n");
                print("\n");
                c = 0;
                start = 2;
            }
            if (start == 2) {
                print("1 2 3 4 5 6 7 8 9 10\n");
                print("X X X X X X X X X X\n");
                print("\n");
                for (x = 1; x <= 10; x++)
                    as[x] = "X";
                start = 0;
            }
            print("INPUT THE NUMBER");
            while (1) {
                n = parseInt(await input());
                if (n >= 0 && n <= 11)
                    break;
                print("ILLEGAL ENTRY--TRY AGAIN.\n");
            }
            if (n == 11) {
                start = 1;
                continue;
            }
            if (n == 0) {
                start = 2;
                continue;
            }
            if (m != n) {
                m = n;
                as[n] = (as[n] == "O" ? "X" : "O");
                do {
                    r = Math.tan(q + n / q - n) - Math.sin(q / n) + 336 * Math.sin(8 * n);
                    n = r - Math.floor(r);
                    n = Math.floor(10 * n);
                    as[n] = (as[n] == "O" ? "X" : "O");
                } while (m == n) ;
            } else {
                as[n] = (as[n] == "O" ? "X" : "O");
                do {
                    r = 0.592 * (1 / Math.tan(q / n + q)) / Math.sin(n * 2 + q) - Math.cos(n);
                    n = r - Math.floor(r);
                    n = Math.floor(10 * n);
                    as[n] = (as[n] == "O" ? "X" : "O");
                } while (m == n) ;
            }
            print("1 2 3 4 5 6 7 8 9 10\n");
            for (z = 1; z <= 10; z++)
                print(as[z] + " ");
            c++;
            print("\n");
            for (z = 1; z <= 10; z++) {
                if (as[z] != "O")
                    break;
            }
        } while (z <= 10) ;
        if (c <= 12) {
            print("VERY GOOD.  YOU GUESSED IT IN ONLY " + c + " GUESSES.\n");
        } else {
            print("TRY HARDER NEXT TIME.  IT TOOK YOU " + c + " GUESSES.\n");
        }
        print("DO YOU WANT TO TRY ANOTHER PUZZLE");
        str = await input();
        if (str.substr(0, 1) == "N")
            break;
    }
    print("\n");
}

main();

File: 36_Flip_Flop/java/FlipFlop.java
import java.util.Scanner;
import java.lang.Math;

/**
 * Game of FlipFlop
 * <p>
 * Based on the BASIC game of FlipFlop here
 * https://github.com/coding-horror/basic-computer-games/blob/main/36%20Flip%20Flop/flipflop.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's BASIC game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 *
 * Converted from BASIC to Java by Darren Cardenas.
 */

public class FlipFlop {

  private final Scanner scan;  // For user input

  private enum Step {
    RANDOMIZE, INIT_BOARD, GET_NUMBER, ILLEGAL_ENTRY, FLIP_POSITION, SET_X_FIRST, SET_X_SECOND,
    GENERATE_R_FIRST, GENERATE_R_SECOND, PRINT_BOARD, QUERY_RETRY
  }

  public FlipFlop() {

    scan = new Scanner(System.in);

  }  // End of constructor FlipFlop

  public void play() {

    showIntro();
    startGame();

  }  // End of method play

  private static void showIntro() {

    System.out.println(" ".repeat(31) + "FLIPFLOP");
    System.out.println(" ".repeat(14) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
    System.out.println("");

  }  // End of method showIntro

  private void startGame() {

    double mathVal = 0;
    double randVal = 0;
    double tmpVal = 0;

    int index = 0;
    int match = 0;
    int numFlip = 0;
    int numGuesses = 0;

    Step nextStep = Step.RANDOMIZE;

    String userResponse = "";

    String[] board = new String[21];

    System.out.println("THE OBJECT OF THIS PUZZLE IS TO CHANGE THIS:");
    System.out.println("");
    System.out.println("X X X X X X X X X X");
    System.out.println("");
    System.out.println("TO THIS:");
    System.out.println("");
    System.out.println("O O O O O O O O O O");
    System.out.println("");
    System.out.println("BY TYPING THE NUMBER CORRESPONDING TO THE POSITION OF THE");
    System.out.println("LETTER ON SOME NUMBERS, ONE POSITION WILL CHANGE, ON");
    System.out.println("OTHERS, TWO WILL CHANGE.  TO RESET LINE TO ALL X'S, TYPE 0");
    System.out.println("(ZERO) AND TO START OVER IN THE MIDDLE OF A GAME, TYPE ");
    System.out.println("11 (ELEVEN).");
    System.out.println("");

    // Begin outer while loop
    while (true) {

      // Begin switch
      switch (nextStep) {

        case RANDOMIZE:

          randVal = Math.random();

          System.out.println("HERE IS THE STARTING LINE OF X'S.");
          System.out.println("");

          numGuesses = 0;
          nextStep = Step.INIT_BOARD;
          break;

        case INIT_BOARD:

          System.out.println("1 2 3 4 5 6 7 8 9 10");
          System.out.println("X X X X X X X X X X");
          System.out.println("");

          // Avoid out of bounds error by starting at zero
          for (index = 0; index <= 10; index++) {
            board[index] = "X";
          }

          nextStep = Step.GET_NUMBER;
          break;

        case GET_NUMBER:

          System.out.print("INPUT THE NUMBER? ");
          userResponse = scan.nextLine();

          try {
            numFlip = Integer.parseInt(userResponse);
          }
          catch (NumberFormatException ex) {
            nextStep = Step.ILLEGAL_ENTRY;
            break;
          }

          // Command to start a new game
          if (numFlip == 11) {
            nextStep = Step.RANDOMIZE;
            break;
          }

          if (numFlip > 11) {
            nextStep = Step.ILLEGAL_ENTRY;
            break;
          }

          // Command to reset the board
          if (numFlip == 0) {
            nextStep = Step.INIT_BOARD;
            break;
          }

          if (match == numFlip) {
            nextStep = Step.FLIP_POSITION;
            break;
          }

          match = numFlip;

          if (board[numFlip].equals("O")) {
            nextStep = Step.SET_X_FIRST;
            break;
          }

          board[numFlip] = "O";
          nextStep = Step.GENERATE_R_FIRST;
          break;

        case ILLEGAL_ENTRY:
          System.out.println("ILLEGAL ENTRY--TRY AGAIN.");
          nextStep = Step.GET_NUMBER;
          break;

        case GENERATE_R_FIRST:

          mathVal = Math.tan(randVal + numFlip / randVal - numFlip) - Math.sin(randVal / numFlip) + 336
                    * Math.sin(8 * numFlip);

          tmpVal = mathVal - (int)Math.floor(mathVal);

          numFlip = (int)(10 * tmpVal);

          if (board[numFlip].equals("O")) {
            nextStep = Step.SET_X_FIRST;
            break;
          }

          board[numFlip] = "O";
          nextStep = Step.PRINT_BOARD;
          break;

        case SET_X_FIRST:
          board[numFlip] = "X";

          if (match == numFlip) {
            nextStep = Step.GENERATE_R_FIRST;
          } else {
            nextStep = Step.PRINT_BOARD;
          }
          break;

        case FLIP_POSITION:

          if (board[numFlip].equals("O")) {
            nextStep = Step.SET_X_SECOND;
            break;
          }

          board[numFlip] = "O";
          nextStep = Step.GENERATE_R_SECOND;
          break;

        case GENERATE_R_SECOND:

          mathVal = 0.592 * (1 / Math.tan(randVal / numFlip + randVal)) / Math.sin(numFlip * 2 + randVal)
                    - Math.cos(numFlip);

          tmpVal = mathVal - (int)mathVal;
          numFlip = (int)(10 * tmpVal);

          if (board[numFlip].equals("O")) {
            nextStep = Step.SET_X_SECOND;
            break;
          }

          board[numFlip] = "O";
          nextStep = Step.PRINT_BOARD;
          break;

        case SET_X_SECOND:

          board[numFlip] = "X";
          if (match == numFlip) {
            nextStep = Step.GENERATE_R_SECOND;
            break;
          }

          nextStep = Step.PRINT_BOARD;
          break;

        case PRINT_BOARD:
          System.out.println("1 2 3 4 5 6 7 8 9 10");

          for (index = 1; index <= 10; index++) {
            System.out.print(board[index] + " ");
          }

          numGuesses++;

          System.out.println("");

          for (index = 1; index <= 10; index++) {
            if (!board[index].equals("O")) {
              nextStep = Step.GET_NUMBER;
              break;
            }
          }

          if (nextStep == Step.GET_NUMBER) {
            break;
          }

          if (numGuesses > 12) {
            System.out.println("TRY HARDER NEXT TIME.  IT TOOK YOU " + numGuesses + " GUESSES.");
          } else {
            System.out.println("VERY GOOD.  YOU GUESSED IT IN ONLY " + numGuesses + " GUESSES.");
          }
          nextStep = Step.QUERY_RETRY;
          break;

        case QUERY_RETRY:

          System.out.print("DO YOU WANT TO TRY ANOTHER PUZZLE? ");
          userResponse = scan.nextLine();

          if (userResponse.toUpperCase().charAt(0) == 'N') {
            return;
          }
          System.out.println("");
          nextStep = Step.RANDOMIZE;
          break;

        default:
          System.out.println("INVALID STEP");
          nextStep = Step.QUERY_RETRY;
          break;

      }  // End of switch

    }  // End outer while loop

  }  // End of method startGame

  public static void main(String[] args) {

    FlipFlop game = new FlipFlop();
    game.play();

  }  // End of method main

}  // End of class FlipFlop
